### ISSUE TITLE
paste setup sometimes fails to "set nopaste"

### ISSUE THREAD

**Issue by firedev** [neovim/neovim#7994]
- `nvim --version`: 
```
$ nvim --version
NVIM v0.2.3-24-gd9b3ebfed
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/local/Homebrew/Library/Homebrew/shims/super/clang -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -I/tmp/neovim-20171125-11833-1ozpdpl/build/config -I/tmp/neovim-20171125-11833-1ozpdpl/src -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/opt/gettext/include -I/usr/include -I/tmp/neovim-20171125-11833-1ozpdpl/build/src/nvim/auto -I/tmp/neovim-20171125-11833-1ozpdpl/build/include
Compiled by pain@nick.local

Features: +acl +iconv +jemalloc +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/HEAD-d9b3ebf/share/nvim"

Run :checkhealth for more info
```

- Operating system/version: macOS 10.13.3
- Terminal name/version: iTerm Build 3.1.20170717-nightly
- `$TERM`: screen-256color

### Steps to reproduce using `nvim -u NORC`

Can't reproduce yet, it works without the config file.

### Actual behaviour

I was reporting this before but the issue was renamed https://github.com/neovim/neovim/issues/6561

Finally caught it in the wild. Here is what I did:

1. Went to insert mode
2. Pasted something from Alfred clipboard history `Cmd-Shift-V`, `Enter`

Result: 
1. The fragment was inserted with `<Paste>` at the end. 
2. `:verb set et?` → `noexpandtab`

### Expected behaviour

* `expandtab` should stay on.
* There should be no `<Paste>`

**Comment by justinmk** [comment]
This is a known issue (yes, it's quite bothersome). Workaround is to 

    :set nopaste

after pasting.

**Comment by firedev** [comment]
Sorry, which issue is known? What about `expandtab`?

**Comment by justinmk** [comment]
`noexpandtab` is one of many side-effects of `set paste` (see `:help 'paste'`), which is basically what happens when a paste is detected. When the paste ends, Nvim should `:set nopaste`, but sometimes it doesn't.

**Comment by firedev** [comment]
Oh got it, this one was driving me crazy, thanks!

**Comment by languitar** [comment]
What is the reason for this behavior? I also sometimes see this.

**Comment by victoriastuart** [comment]
---

**Update:**

See my answer at https://stackoverflow.com/questions/53123297/disable-insert-paste-mode/53285805#53285805

**TL/DR:**

* disable (comment out or remove) `set nopaste` in ~/.vimrc
* add these to ~/.vimrc:
  * `set clipboard+=unnamed`
  * ```nnoremap p p`]<Esc>```
* use `p` to paste (Normal mode)
* Insert blank line without entering insert mode:
  * above current line: `nmap <A-CR> O<Esc>`
  * below current line: `nmap <CR> o<Esc>`

---

Related discussion:

* https://github.com/neovim/neovim/issues/9203
* https://www.reddit.com/r/neovim/comments/9ufg11/disable_insert_paste_mode/
* https://www.reddit.com/r/vim/comments/9uerhp/disable_insert_paste_mode/
* https://stackoverflow.com/questions/53123297/disable-insert-paste-mode/53285805#53285805

**Comment by frabjous** [comment]
I'm having a version of this problem, where if I use the terminal emulator's paste-from-clipboard short key (Ctrl-Shift-V for me) when in the command line in neovim, it'll then put me in -- INSERT (paste) -- mode and keep me in it until I manually `set nopaste`.
Even the workaround does not work for me. Neovim reports nopaste as being set, e.g., if I type in `set paste?`, but I always get INSERT (paste) mode anyway.
